Compilation Error from the latest Source Code in GitHub possibly caused by CUSPARESE APIs

Kindly need help. 
I'm currently trying to install cupy packages by compiling the source code released in the GitHub.

My System version is:
`Ubuntu 18.04`

My CUDA version is:
`Copyright (c) 2005-2019 NVIDIA Corporation`
`Built on Fri_Feb__8_19:08:17_PST_2019`
`Cuda compilation tools, release 10.1, V10.1.105`

My Python version is:
`Python 3.8.5`

And when installing the packages by the following commands:
`cd cupy`
`pip install . -vvvv`

Some errors occur:
`In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:`
`cupy_backends/cuda/libs/../cupy_cusparse.h:282:17: error: conflicting declaration ‘typedef enum cusparseIndexType_t cusparseIndexType_t’`
`typedef enum {} cusparseIndexType_t`
 `                             ^~~~~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:`
  `/usr/local/cuda-10.1/include/cusparse.h:6960:3: note: previous declaration as ‘typedef enum cusparseIndexType_t cusparseIndexType_t’
   } cusparseIndexType_t;
     ^~~~~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:283:17: error: conflicting declaration ‘typedef enum cusparseFormat_t cusparseFormat_t’
   typedef enum {} cusparseFormat_t;
                   ^~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6944:3: note: previous declaration as ‘typedef enum cusparseFormat_t cusparseFormat_t’
   } cusparseFormat_t;
     ^~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:284:17: error: conflicting declaration ‘typedef enum cusparseOrder_t cusparseOrder_t’
   typedef enum {} cusparseOrder_t;
                   ^~~~~~~~~~~~~~~`
  `In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6949:3: note: previous declaration as ‘typedef enum cusparseOrder_t cusparseOrder_t’
   } cusparseOrder_t;
     ^~~~~~~~~~~~~~~`
  `In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:286:17: error: conflicting declaration ‘typedef enum cusparseSpMMAlg_t cusparseSpMMAlg_t’
   typedef enum {} cusparseSpMMAlg_t;
                   ^~~~~~~~~~~~~~~~~`
`  In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6955:3: note: previous declaration as ‘typedef enum cusparseSpMMAlg_t cusparseSpMMAlg_t’
   } cusparseSpMMAlg_t;
     ^~~~~~~~~~~~~~~~~`
`  In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:450:17: error: conflicting declaration ‘typedef enum cusparseCsr2CscAlg_t cusparseCsr2CscAlg_t’
   typedef enum {} cusparseCsr2CscAlg_t;
                   ^~~~~~~~~~~~~~~~~~~~`
`  In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6898:3: note: previous declaration as ‘typedef enum cusparseCsr2CscAlg_t cusparseCsr2CscAlg_t’
   } cusparseCsr2CscAlg_t;
     ^~~~~~~~~~~~~~~~~~~~`


I search for several ways attempting to solve the conflicts, including finding the declarations in the `cusparse.h` included in CUDA-10,1 and trying to make correspondences between cupy's and CUDA's conflicted declarations. However, no one works. So I am sincerely asking for some help.

By the way, I have already tried to install by `pip install --pre cupy-cuda101`. There are also some errors in the linking reference of cupy to 'libcusparse.so', with the same familiar functions. I suspect there may be relationships between the compilation problems and the linking problems.

--- a/cupy_backends/cuda/cupy_cusparse.h
+++ b/cupy_backends/cuda/cupy_cusparse.h
@@ -5,176 +5,430 @@
 #include <cusparse.h>
 
 #if !defined(CUSPARSE_VERSION)
-#if CUDA_VERSION < 10000
-#define CUSPARSE_VERSION CUDA_VERSION // CUDA_VERSION used instead
+  // CUSPARSE_VERSION introduced in CUDA 10.1 Update 2 (10.1.243).
+  #if CUDA_VERSION < 10000
+    #define CUSPARSE_VERSION CUDA_VERSION // CUDA_VERSION used instead
+  #elif CUDA_VERSION < 10010  // CUDA 10.0.x
+    # define CUSPARSE_VERSION 10000
+  #elif CUDA_VERSION < 10020  // CUDA 10.1.x
+    // CUSPARSE_VER_MAJOR introduced in CUDA 10.1 Update 1 (10.1.168).
+    #if !defined(CUSPARSE_VER_MAJOR)
+      // CUDA 10.1 (10.1.105) contains cuSPARSE 10.1
+      #define CUSPARSE_VERSION 10010
+    #else
+      // CUDA 10.1 Update 1 (10.1.168) contains cuSPARSE 10.2.0.0
+      #define CUSPARSE_VERSION (CUSPARSE_VER_MAJOR * 1000 + \
+                                CUSPARSE_VER_MINOR *  100 + \
+                                CUSPARSE_VER_PATCH)
+    #endif
+  #endif
+#endif  // #if !defined(CUSPARSE_VERSION)
+
+/*
+ * Generic APIs are not suppoted in CUDA 10.1/10.2 on Windows.
+ * These APIs are available in headers in CUDA 10.1 and 10.1 Update 1, but hidden in
+ * 10.1 Update 2 and 10.2 on Windows. Implementations are not exposed from DLL in either
+ * versions of CUDA, so we have to provide a stub implementation using the full signature
+ * declared in the header, instead of using (...).
+ */
+
+#if defined(_WIN32) && (CUSPARSE_VERSION < 11000)
+  #if 10200 < CUSPARSE_VERSION
+    #define WIN32_EXPOSE_SPMM_STUB_DECL 1
+  #else
+    #define WIN32_EXPOSE_SPMM_STUB_DECL 0
+  #endif
+  #define WIN32_EXPOSE_SPMM_STUB_IMPL 1
 #else
-#define CUSPARSE_VERSION 10000
+  #define WIN32_EXPOSE_SPMM_STUB_DECL 0
+  #define WIN32_EXPOSE_SPMM_STUB_IMPL 0
 #endif
-#endif // #if !defined(CUSPARSE_VERSION)
 
-#if (CUSPARSE_VERSION < 10200) || (CUSPARSE_VERSION < 11000 && defined(_WIN32))
-// Types, macro and functions added in cuSparse 10.2
-// Windows support added in cuSparse 11.0
+#if CUSPARSE_VERSION < 10010
+// Added in cuSPARSE 10.1 (CUDA 10.1.105)
 
-// cuSPARSE generic API
-typedef void* cusparseSpVecDescr_t;
-typedef void* cusparseDnVecDescr_t;
+// CSR2CSC
+typedef enum {} cusparseCsr2CscAlg_t;
+
+#endif  // #if CUSPARSE_VERSION < 10010
+
+#if CUSPARSE_VERSION < 10010 || WIN32_EXPOSE_SPMM_STUB_DECL
+// Generic API types added in cuSPARSE 10.1 (CUDA 10.1.105)
+
 typedef void* cusparseSpMatDescr_t;
 typedef void* cusparseDnMatDescr_t;
-
 typedef enum {} cusparseIndexType_t;
 typedef enum {} cusparseFormat_t;
 typedef enum {} cusparseOrder_t;
-typedef enum {} cusparseSpMVAlg_t;
 typedef enum {} cusparseSpMMAlg_t;
 
-cusparseStatus_t cusparseCreateSpVec(...) {
+#endif  // #if CUSPARSE_VERSION < 10010 || WIN32_EXPOSE_SPMM_STUB_DECL
+
+#if CUSPARSE_VERSION < 10200
+// Added in cuSPARSE 10.2 (CUDA 10.1.168)
+
+// CSR2CSC
+cusparseStatus_t cusparseCsr2cscEx2_bufferSize(...) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDestroySpVec(...) {
+cusparseStatus_t cusparseCsr2cscEx2(...) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVecGet(...) {
+#endif  // #if CUSPARSE_VERSION < 10200
+
+#if CUSPARSE_VERSION < 10200 || WIN32_EXPOSE_SPMM_STUB_DECL
+// Generic API types added in cuSPARSE 10.2 (CUDA 10.1.168)
+
+typedef void* cusparseSpVecDescr_t;
+typedef void* cusparseDnVecDescr_t;
+typedef enum {} cusparseSpMVAlg_t;
+
+#endif  // #if CUSPARSE_VERSION < 10200 || WIN32_EXPOSE_SPMM_STUB_DECL
+
+#if CUSPARSE_VERSION < 10200 || WIN32_EXPOSE_SPMM_STUB_IMPL
+// Generic APIs added in cuSPARSE 10.2 (CUDA 10.1.168)
+
+cusparseStatus_t cusparseCreateSpVec(
+    cusparseSpVecDescr_t* spVecDescr,
+    int64_t               size,
+    int64_t               nnz,
+    void*                 indices,
+    void*                 values,
+    cusparseIndexType_t   idxType,
+    cusparseIndexBase_t   idxBase,
+    cudaDataType          valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVecGetIndexBase(...) {
+cusparseStatus_t cusparseDestroySpVec(cusparseSpVecDescr_t spVecDescr) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVecGetValues(...) {
+cusparseStatus_t cusparseSpVecGet(
+    const cusparseSpVecDescr_t spVecDescr,
+    int64_t*                   size,
+    int64_t*                   nnz,
+    void**                     indices,
+    void**                     values,
+    cusparseIndexType_t*       idxType,
+    cusparseIndexBase_t*       idxBase,
+    cudaDataType*              valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVecSetValues(...) {
+cusparseStatus_t cusparseSpVecGetIndexBase(
+    const cusparseSpVecDescr_t spVecDescr,
+    cusparseIndexBase_t*       idxBase) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCreateCoo(...) {
+cusparseStatus_t cusparseSpVecGetValues(
+    const cusparseSpVecDescr_t spVecDescr,
+    void**                     values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCreateCooAoS(...) {
+cusparseStatus_t cusparseSpVecSetValues(
+    cusparseSpVecDescr_t spVecDescr,
+    void*                values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCreateCsr(...) {
+cusparseStatus_t cusparseCreateCoo(
+    cusparseSpMatDescr_t* spMatDescr,
+    int64_t               rows,
+    int64_t               cols,
+    int64_t               nnz,
+    void*                 cooRowInd,
+    void*                 cooColInd,
+    void*                 cooValues,
+    cusparseIndexType_t   cooIdxType,
+    cusparseIndexBase_t   idxBase,
+    cudaDataType          valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDestroySpMat(...) {
+cusparseStatus_t cusparseCreateCooAoS(
+    cusparseSpMatDescr_t* spMatDescr,
+    int64_t               rows,
+    int64_t               cols,
+    int64_t               nnz,
+    void*                 cooInd,
+    void*                 cooValues,
+    cusparseIndexType_t   cooIdxType,
+    cusparseIndexBase_t   idxBase,
+    cudaDataType          valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCooGet(...) {
+
+cusparseStatus_t cusparseCreateCsr(
+    cusparseSpMatDescr_t* spMatDescr,
+    int64_t               rows,
+    int64_t               cols,
+    int64_t               nnz,
+    void*                 csrRowOffsets,
+    void*                 csrColInd,
+    void*                 csrValues,
+    cusparseIndexType_t   csrRowOffsetsType,
+    cusparseIndexType_t   csrColIndType,
+    cusparseIndexBase_t   idxBase,
+    cudaDataType          valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCooAoSGet(...) {
+cusparseStatus_t cusparseDestroySpMat(cusparseSpMatDescr_t spMatDescr) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCsrGet(...) {
+cusparseStatus_t cusparseCooGet(
+    const cusparseSpMatDescr_t spMatDescr,
+    int64_t*                   rows,
+    int64_t*                   cols,
+    int64_t*                   nnz,
+    void**                     cooRowInd,  // COO row indices
+    void**                     cooColInd,  // COO column indices
+    void**                     cooValues,  // COO values
+    cusparseIndexType_t*       idxType,
+    cusparseIndexBase_t*       idxBase,
+    cudaDataType*              valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMatGetFormat(...) {
+cusparseStatus_t cusparseCooAoSGet(
+    const cusparseSpMatDescr_t spMatDescr,
+    int64_t*                   rows,
+    int64_t*                   cols,
+    int64_t*                   nnz,
+    void**                     cooInd,     // COO indices
+    void**                     cooValues,  // COO values
+    cusparseIndexType_t*       idxType,
+    cusparseIndexBase_t*       idxBase,
+    cudaDataType*              valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMatGetIndexBase(...) {
+cusparseStatus_t cusparseCsrGet(
+    const cusparseSpMatDescr_t spMatDescr,
+    int64_t*                   rows,
+    int64_t*                   cols,
+    int64_t*                   nnz,
+    void**                     csrRowOffsets,
+    void**                     csrColInd,
+    void**                     csrValues,
+    cusparseIndexType_t*       csrRowOffsetsType,
+    cusparseIndexType_t*       csrColIndType,
+    cusparseIndexBase_t*       idxBase,
+    cudaDataType*              valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMatGetValues(...) {
+cusparseStatus_t cusparseSpMatGetFormat(
+    const cusparseSpMatDescr_t spMatDescr,
+    cusparseFormat_t*          format) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMatSetValues(...) {
+cusparseStatus_t cusparseSpMatGetIndexBase(
+    const cusparseSpMatDescr_t spMatDescr,
+    cusparseIndexBase_t*       idxBase) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMatGetStridedBatch(...) {
+cusparseStatus_t cusparseSpMatGetValues(
+    const cusparseSpMatDescr_t spMatDescr,
+    void**                     values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMatSetStridedBatch(...) {
+cusparseStatus_t cusparseSpMatSetValues(
+    cusparseSpMatDescr_t spMatDescr,
+    void*                values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCreateDnVec(...) {
+cusparseStatus_t cusparseSpMatGetStridedBatch(
+    const cusparseSpMatDescr_t spMatDescr,
+    int*                       batchCount) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDestroyDnVec(...) {
+cusparseStatus_t cusparseSpMatSetStridedBatch(
+    cusparseSpMatDescr_t spMatDescr,
+    int                  batchCount) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnVecGet(...) {
+cusparseStatus_t cusparseCreateDnVec(
+    cusparseDnVecDescr_t* dnVecDescr,
+    int64_t               size,
+    void*                 values,
+    cudaDataType          valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnVecGetValues(...) {
+cusparseStatus_t cusparseDestroyDnVec(cusparseDnVecDescr_t dnVecDescr) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnVecSetValues(...) {
+cusparseStatus_t cusparseDnVecGet(
+    const cusparseDnVecDescr_t dnVecDescr,
+    int64_t*                   size,
+    void**                     values,
+    cudaDataType*              valueType) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseCreateDnMat(...) {
+cusparseStatus_t cusparseDnVecGetValues(
+    const cusparseDnVecDescr_t dnVecDescr,
+    void**                     values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDestroyDnMat(...) {
+cusparseStatus_t cusparseDnVecSetValues(
+    cusparseDnVecDescr_t dnVecDescr,
+    void*                values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnMatGet(...) {
+cusparseStatus_t cusparseCreateDnMat(
+    cusparseDnMatDescr_t* dnMatDescr,
+    int64_t               rows,
+    int64_t               cols,
+    int64_t               ld,
+    void*                 values,
+    cudaDataType          valueType,
+    cusparseOrder_t       order) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnMatGetValues(...) {
+cusparseStatus_t cusparseDestroyDnMat(cusparseDnMatDescr_t dnMatDescr) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnMatSetValues(...) {
+cusparseStatus_t cusparseDnMatGet(
+    const cusparseDnMatDescr_t dnMatDescr,
+    int64_t*                   rows,
+    int64_t*                   cols,
+    int64_t*                   ld,
+    void**                     values,
+    cudaDataType*              type,
+    cusparseOrder_t*           order) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnMatGetStridedBatch(...) {
+cusparseStatus_t cusparseDnMatGetValues(
+    const cusparseDnMatDescr_t dnMatDescr,
+    void**                     values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseDnMatSetStridedBatch(...) {
+cusparseStatus_t cusparseDnMatSetValues(
+    cusparseDnMatDescr_t dnMatDescr,
+    void*                values) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVV_bufferSize(...) {
+cusparseStatus_t cusparseDnMatGetStridedBatch(
+    const cusparseDnMatDescr_t dnMatDescr,
+    int*                       batchCount,
+    int64_t*                   batchStride) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVV(...) {
+cusparseStatus_t cusparseDnMatSetStridedBatch(
+    cusparseDnMatDescr_t dnMatDescr,
+    int                  batchCount,
+    int64_t              batchStride) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMV_bufferSize(...) {
+cusparseStatus_t cusparseSpVV_bufferSize(
+    cusparseHandle_t           handle,
+    cusparseOperation_t        opX,
+    const cusparseSpVecDescr_t vecX,
+    const cusparseDnVecDescr_t vecY,
+    const void*                result,
+    cudaDataType               computeType,
+    size_t*                    bufferSize) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMV(...) {
+cusparseStatus_t cusparseSpVV(
+    cusparseHandle_t           handle,
+    cusparseOperation_t        opX,
+    const cusparseSpVecDescr_t vecX,
+    const cusparseDnVecDescr_t vecY,
+    void*                      result,
+    cudaDataType               computeType,
+    void*                      externalBuffer) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMM_bufferSize(...) {
+cusparseStatus_t cusparseSpMV_bufferSize(
+    cusparseHandle_t           handle,
+    cusparseOperation_t        opA,
+    const void*                alpha,
+    const cusparseSpMatDescr_t matA,
+    const cusparseDnVecDescr_t vecX,
+    const void*                beta,
+    const cusparseDnVecDescr_t vecY,
+    cudaDataType               computeType,
+    cusparseSpMVAlg_t          alg,
+    size_t*                    bufferSize) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpMM(...) {
+cusparseStatus_t cusparseSpMV(
+    cusparseHandle_t           handle,
+    cusparseOperation_t        opA,
+    const void*                alpha,
+    const cusparseSpMatDescr_t matA,
+    const cusparseDnVecDescr_t vecX,
+    const void*                beta,
+    const cusparseDnVecDescr_t vecY,
+    cudaDataType               computeType,
+    cusparseSpMVAlg_t          alg,
+    void*                      externalBuffer) {
   return CUSPARSE_STATUS_SUCCESS;
 }
+
+cusparseStatus_t cusparseSpMM_bufferSize(
+    cusparseHandle_t           handle,
+    cusparseOperation_t        opA,
+    cusparseOperation_t        opB,
+    const void*                alpha,
+    const cusparseSpMatDescr_t matA,
+    const cusparseDnMatDescr_t matB,
+    const void*                beta,
+    cusparseDnMatDescr_t       matC,
+    cudaDataType               computeType,
+    cusparseSpMMAlg_t          alg,
+    size_t*                    bufferSize) {
+  return CUSPARSE_STATUS_SUCCESS;
+}
+
+cusparseStatus_t cusparseSpMM(
+    cusparseHandle_t           handle,
+    cusparseOperation_t        opA,
+    cusparseOperation_t        opB,
+    const void*                alpha,
+    const cusparseSpMatDescr_t matA,
+    const cusparseDnMatDescr_t matB,
+    const void*                beta,
+    cusparseDnMatDescr_t       matC,
+    cudaDataType               computeType,
+    cusparseSpMMAlg_t          alg,
+    void*                      externalBuffer) {
+  return CUSPARSE_STATUS_SUCCESS;
+}
+
+#endif  // #if CUSPARSE_VERSION < 10200 || WIN32_EXPOSE_SPMM_STUB_IMPL
+
+#if CUSPARSE_VERSION < 10300 || WIN32_EXPOSE_SPMM_STUB_IMPL
+// Generic APIs added in cuSPARSE 10.3 (CUDA 10.1.243)
 
 cusparseStatus_t cusparseConstrainedGeMM_bufferSize(...) {
   return CUSPARSE_STATUS_SUCCESS;
@@ -184,22 +438,7 @@ cusparseStatus_t cusparseConstrainedGeMM(...) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-#endif // #if CUSPARSE_VERSION < 10200
-
-#if CUSPARSE_VERSION < 10200
-// Functions added in cuSparse 10.2
-
-// CSR2CSC
-typedef enum {} cusparseCsr2CscAlg_t;
-
-cusparseStatus_t cusparseCsr2cscEx2_bufferSize(...) {
-  return CUSPARSE_STATUS_SUCCESS;
-}
-
-cusparseStatus_t cusparseCsr2cscEx2(...) {
-  return CUSPARSE_STATUS_SUCCESS;
-}
-#endif // #if CUSPARSE_VERSION < 10200
+#endif  // #if CUSPARSE_VERSION < 10300 || WIN32_EXPOSE_SPMM_STUB_IMPL
 
 #if CUSPARSE_VERSION >= 11000
 // Functions deleted in cuSparse 11.0

--- a/cupy_backends/cuda/cupy_cusparse.h
+++ b/cupy_backends/cuda/cupy_cusparse.h
@@ -26,10 +26,8 @@
 
 /*
  * Generic APIs are not suppoted in CUDA 10.1/10.2 on Windows.
- * These APIs are available in headers in CUDA 10.1 and 10.1 Update 1, but hidden in
- * 10.1 Update 2 and 10.2 on Windows. Implementations are not exposed from DLL in either
- * versions of CUDA, so we have to provide a stub implementation using the full signature
- * declared in the header, instead of using (...).
+ * These APIs are available in headers in CUDA 10.1 and 10.1 Update 1,
+ * but hidden in 10.1 Update 2 and 10.2 on Windows.
  */
 
 #if defined(_WIN32) && (CUSPARSE_VERSION < 11000)
@@ -89,6 +87,13 @@ typedef enum {} cusparseSpMVAlg_t;
 
 #if CUSPARSE_VERSION < 10200 || WIN32_EXPOSE_SPMM_STUB_IMPL
 // Generic APIs added in cuSPARSE 10.2 (CUDA 10.1.168)
+
+/*
+ * On Windows, implementations are not exposed from DLL in CUDA 10.1/10.2
+ * although it is declared in the header. So we have to provide a stub
+ * implementation using the full signature to match with the signature
+ * declared in cusparse.h, instead of using (...).
+ */
 
 cusparseStatus_t cusparseCreateSpVec(
     cusparseSpVecDescr_t* spVecDescr,

